rust/core-lib/src/view.rs

```diff
@@ -612,28 +612,40 @@ impl View {
         !self.selection.regions_in_range(offset, offset).is_empty()
     }
 
-    // Render a single line, and advance cursors to next line.
+    // Encode a single line with its styles and cursors in JSON.
+    // If "text" is not specified, don't add "text" to the output.
+    // If "style_spans" are not specified, don't add "styles" to the output.
     fn render_line(
         &self,
         client: &Client,
         styles: &StyleMap,
-        text: &Rope,
         line: VisualLine,
-        style_spans: &Spans<Style>,
-        line_num: usize,
+        text: Option<&Rope>,
+        style_spans: Option<&Spans<Style>>,
     ) -> Value {
         let start_pos = line.interval.start;
         let pos = line.interval.end;
-        let l_str = text.slice_to_cow(start_pos..pos);
         let mut cursors = Vec::new();
         let mut selections = Vec::new();
         for region in self.selection.regions_in_range(start_pos, pos) {
             // cursor
             let c = region.end;
             if (c > start_pos && c < pos)
                 || (!region.is_upstream() && c == start_pos)
-                || (region.is_upstream() && c == pos)
-                || (c == pos && c == text.len() && self.line_of_offset(text, c) == line_num)
+
+            // TODO: Most likely, "(region.is_upstream() && c == pos)"
+            // is an error. The correct action here is to compare region.start
+            // (not region.end) and affinity against the interval's end point.
+            //
+            //  || (region.is_upstream() && c == pos)
+                || (region.is_upstream() && pos == region.start)
+            // TODO: Not clear why this case is special.
+            // The line overlap with the region should be 100% determined by
+            // region start/end, region affinity, and line start/end.
+            // No special cases here and no dependency on "text".
+            // Comment out the condition for now.
+            //
+            //  || (c == pos && c == text.len() && self.line_of_offset(text, c) == line_num)
```

Comment:
Have you talked about these two with @cmyr or @raphlinus ? I have to admit I don't know the inner workings of xi too well, so not sure if this doesn't have unintended effects

Reply:
AFAIK cmyr is busy and he cannot participate in code reviews.
This if clause is written by Raph but looks like he isn't coming in the comments.
I will try PMing him or notifying him on zulip.

I tested my whole change in a real xi-term session and didn't see any issues with moving over EOLs (which is what is covered by this clause). But there is still a chance that I'm missing something.

Reply:
OK, Raph provided some explanations on zulip and I also gave a second look on selection.rs myself. Bottom line: these checks are legal (save for a minor issue: can remove the line_of_offset check as it always passes) and the last one is indeed a special case.

So I'm taking back my edits here, and will not change the behavior of interval matching in render_line(). I'll only make a few cosmetic changes.

Reply:
Okie :+1: